### DISCUSSION TITLE
[ 功能建议 ]  数据库工单划分类别, 划分为上线工单和业务工单

### DISCUSSION THREAD

**Discussion by ruanct** [hhyo/Archery#1586] [Ideas]
### 问题描述

由于老旧系统(基于Oracle)的设计问题，目前存在一些异常数据修改的操作，这样实际就造成了工单有两个分类，一为系统上线工单，二为业务数据工单。

系统上的工单还没有类型划分，是否可以考虑提供一个工单类型属性，用以区别不同类型工单。
这样在工单查询结果页面，dba可以根据工单类型做筛选查询，便于调阅该类工单历史。

好人好运，感谢各位大神的付出与支持，祝好！

**Comment by ruanct** [comment]
写错版面了，应该写在功能建议版， 见谅

**Comment by LeoQuote** [comment]
我想你的意思应该是类似github issue 这样, 可以给工单打上标签, 这样有利于检索和审查, 确实是一个很好的想法.

但是目前工单已经有简陋的搜索功能, 完全可以要求在提交工单时标题加上对应的关键字, 你看这个能不能满足一些你的需求? 

如果这个不能满足的话, 还可以把archery 自己的表接入archery , 在查询页面使用sql语句进行查询, 包括归集, 统计之类的都可以做到, 还可以导出到excel 中.

**Comment by ruanct** [comment]
感谢 LeoQuote，您提议的方法，我认为应该是一个过渡、折中的办法。
实际上，我这边mysql生产环境已经用起来这套系统，随着时间推移，业务工单越来越多，涉及的实例越来越多。工单查询的过滤条件除了工单状态，我认为应该再加上涉及实例(a)、工单类型(b)，甚至加上工单提出部门或者组(c)。

现阶段可以按您提议的要求来。祝好！

**Comment by LeoQuote** [comment]
如果是涉及实例, 工单类型, 工单提出部门等已存在的字段的话, 这个功能应该还是比较好做的, 只是不知道要怎么做, 类似于kibana那样的一个个的过滤器吗 @hhyo

**Comment by hhyo** [comment]
增加筛选项是简单的，参考实例列表，不过现在的filter实现的太烂，会生成复杂子查询，可以改为 `objects.filter(**filter_dict)` 

工单筛选的用途是什么，是为了用于统计呢还是检索方便，统计的话直接出几个维度的报表更合适，检索的话还是搜索匹配更合适，不过多加几个筛选总是好的😁

**Comment by hhyo** [comment]
@LeoQuote  筛选项的样例提了PR #288  测试还没写~~

不过@ericruan-cn 的需求是新增一个工单类型，这个类型由发起人选择？

**Comment by ruanct** [comment]
@hhyo 有一个折中办法，就是不增加类型，而是工单创建页面提供一个工单前缀选择(1)系统上线(2)业务数据。 当选择系统上线选项， 只是在最后生成的工单名称强制加上前缀 “系统上线_”. 这样产生的工单就可以根据前缀来检索。 这种思路和操作应该能够接受。

最友好的工单名称应该类似这种:
系统上线_XX系统_XXXXXX
数据变更_XX系统_XXXXXX

这个需求属于锦上添花类型，目前也能参照 @leiQ的折中办法，可以再想想

**Comment by hhyo** [comment]
加类型是很好的，只是这个要通用化，最好是可以由管理员来定义

**Comment by hhyo** [comment]
- 增加系统字典表，允许管理员在字典表中定义和维护工单类型
- 提交工单时可以下拉选择工单类型
- 工单列表增加工单类型筛选项

**Reply by carllhw** [reply]
工单类型功能会实现吗

**Comment by carllhw** [comment]
工单类型功能会实现吗

**Reply by LeoQuote** [reply]
你好, 本项目是一个开源项目, 你可以自由提功能需求, 但其实比提功能需求更加务实, 有价值的是实现这个功能, 本项目开发文档如下, 希望对你有帮助:

* https://github.com/hhyo/Archery/wiki/development
* https://github.com/hhyo/Archery/wiki/notifier

期待你的pr, 谢谢~
